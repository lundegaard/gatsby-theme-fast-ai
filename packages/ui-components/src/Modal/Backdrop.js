import React from 'react';
import PropTypes from 'prop-types';

import Box from '../Box';

const Backdrop = ({ sx, ...rest }) => (
	<Box
		sx={{
			position: 'fixed',
			top: 0,
			left: 0,
			right: 0,
			bottom: 0,
			backgroundColor: 'body',
			opacity: 0.5,
			zIndex: 1000,
			...sx,
		}}
		{...rest}
	/>
);
Backdrop.propTypes = { sx: PropTypes.object };

export default Backdrop;
