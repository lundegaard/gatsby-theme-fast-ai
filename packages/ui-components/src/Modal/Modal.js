import React from 'react';
import PropTypes from 'prop-types';

import Box from '../Box';

const Modal = ({ children, sx, ...rest }) => (
	<Box
		sx={{
			borderRadius: '5px',
			backgroundColor: 'background',
			width: ['100%', '75%'],
			overflowY: 'auto',
			overflowX: 'hidden',
			WebkitOverflowScrolling: 'touch',
			...sx,
		}}
		{...rest}
	>
		{children}
	</Box>
);

Modal.propTypes = {
	children: PropTypes.node,
	sx: PropTypes.object,
};

export default Modal;
