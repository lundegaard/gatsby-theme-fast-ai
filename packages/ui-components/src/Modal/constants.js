export const MODAL_ROOT = 'modal-root';
