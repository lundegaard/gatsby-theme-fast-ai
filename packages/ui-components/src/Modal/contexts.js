import { createContext } from 'react';

export const ModalApiContext = createContext();
