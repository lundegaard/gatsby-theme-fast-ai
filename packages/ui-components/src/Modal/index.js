export { default } from './Modal';
export { default as ModalProvider } from './ModalProvider';
export { default as ModalRoot } from './ModalRoot';
export * from './constants';
export * from './hooks';
