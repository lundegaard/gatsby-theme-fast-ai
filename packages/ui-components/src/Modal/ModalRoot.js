import React from 'react';

import { MODAL_ROOT } from './constants';

const ModalRoot = () => <div id={MODAL_ROOT} />;

export default ModalRoot;
