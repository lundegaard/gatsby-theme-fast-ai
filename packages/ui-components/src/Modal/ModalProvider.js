import React, { useMemo, useReducer } from 'react';
import PropTypes from 'prop-types';
import { valueMirror } from 'ramda-extension';

import Flex from '../Flex';

import PortalToRoot from './PortalToRoot';
import Backdrop from './Backdrop';
import { ModalApiContext } from './contexts';

const initialState = { modalProps: null, component: null, isOpened: false };
const ActionTypes = valueMirror(['OPEN_MODAL', 'CLOSE_MODAL']);

const reducer = (state, action) => {
	switch (action.type) {
		case ActionTypes.OPEN_MODAL: {
			const { component, modalProps } = action.payload;
			return {
				component,
				modalProps,
				isOpened: true,
			};
		}
		case ActionTypes.CLOSE_MODAL: {
			return {
				component: null,
				modalProps: null,
				isOpened: false,
			};
		}
	}
	return state;
};

const actions = {
	openModal: payload => ({ type: ActionTypes.OPEN_MODAL, payload }),
	closeModal: payload => ({ type: ActionTypes.CLOSE_MODAL, payload }),
};

const toggleBodyClassname = (className, toggler) => {
	if (typeof window === 'undefined') {
		return;
	}
	const classList = document.body.classList;
	return toggler ? classList.add(className) : classList.remove(className);
};

export const ModalProvider = ({ children, bodyModalClassname = 'modal' }) => {
	const [{ isOpened, modalProps, component: ModalComponent }, dispatch] = useReducer(
		reducer,
		initialState
	);

	const api = useMemo(
		() => ({
			openModal: (component, modalProps) => {
				dispatch(actions.openModal({ component, modalProps }));

				toggleBodyClassname(`${bodyModalClassname}--opened`, true);
			},
			closeModal: component => {
				dispatch(actions.closeModal({ component }));

				toggleBodyClassname(`${bodyModalClassname}--opened`, false);
			},
		}),
		[bodyModalClassname]
	);

	return (
		<ModalApiContext.Provider value={api}>
			{children}
			{isOpened && ModalComponent && (
				<PortalToRoot>
					<Backdrop onClick={() => api.closeModal(ModalComponent)} />
					<Flex
						justifyContent="center"
						alignItems="center"
						flexDirection="column"
						sx={{
							zIndex: 1001,
							position: 'fixed',
							top: 0,
							left: 0,
							right: 0,
							bottom: 0,
							p: [0, 4],
						}}
					>
						<ModalComponent {...modalProps} {...api} />
					</Flex>
				</PortalToRoot>
			)}
		</ModalApiContext.Provider>
	);
};

ModalProvider.propTypes = {
	bodyModalClassname: PropTypes.string,
	children: PropTypes.node,
};

export default ModalProvider;
