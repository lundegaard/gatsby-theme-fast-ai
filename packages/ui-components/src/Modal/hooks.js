import { useContext, useMemo } from 'react';

import { ModalApiContext } from './contexts';

export const useModal = ({ component }) => {
	const modalApi = useContext(ModalApiContext);

	const api = useMemo(
		() => ({
			openModal: (modalProps) => {
				modalApi.openModal(component, modalProps);
			},
			closeModal: () => {
				modalApi.closeModal(component);
			},
		}),
		[component, modalApi]
	);

	return api;
};
