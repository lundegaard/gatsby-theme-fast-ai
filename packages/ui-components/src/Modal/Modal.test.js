import React from 'react';
import { shallow } from 'enzyme';

import Modal from './Modal';

describe('<Modal />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Modal {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
