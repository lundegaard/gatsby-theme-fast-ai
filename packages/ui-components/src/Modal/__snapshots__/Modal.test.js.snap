// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> matches snapshot when shallowly rendered 1`] = `
<Box
  sx={
    {
      "WebkitOverflowScrolling": "touch",
      "backgroundColor": "background",
      "borderRadius": "5px",
      "overflowX": "hidden",
      "overflowY": "auto",
      "width": [
        "100%",
        "75%",
      ],
    }
  }
/>
`;
